pkg/reconciler/pipelinerun/resources/resultrefresolution.go

```diff
@@ -109,64 +109,73 @@ func removeDup(refs ResolvedResultRefs) ResolvedResultRefs {
+				return nil, resultRef.PipelineTask, err
+			}
+			resolvedResultRefs = append(resolvedResultRefs, resolved)
 		}
-		resolvedResultRefs = append(resolvedResultRefs, resolved)
 	}
 	return resolvedResultRefs, "", nil
 }
 
-func resolveResultRef(pipelineState PipelineRunState, resultRef *v1.ResultRef) (*ResolvedResultRef, string, error) {
-	referencedPipelineTask := pipelineState.ToMap()[resultRef.PipelineTask]
-	if referencedPipelineTask == nil {
-		return nil, resultRef.PipelineTask, fmt.Errorf("could not find task %q referenced by result", resultRef.PipelineTask)
-	}
-	if !referencedPipelineTask.isSuccessful() && !referencedPipelineTask.isFailure() {
-		return nil, resultRef.PipelineTask, fmt.Errorf("task %q referenced by result was not finished", referencedPipelineTask.PipelineTask.Name)
+func resolveCustomResultRef(customRuns []*v1beta1.CustomRun, resultRef *v1.ResultRef) (*ResolvedResultRef, error) {
+	customRun := customRuns[0]
+	runName := customRun.GetObjectMeta().GetName()
+	runValue, err := findRunResultForParam(customRun, resultRef)
+	if err != nil {
+		return nil, err
 	}
+	return &ResolvedResultRef{
+		Value:           *v1.NewStructuredValues(runValue),
+		FromTaskRun:     "",
+		FromRun:         runName,
+		ResultReference: *resultRef,
+	}, nil
+}
 
-	var runName, runValue, taskRunName string
-	var resultValue v1.ResultValue
-	var err error
-	if referencedPipelineTask.IsCustomTask() {
-		if len(referencedPipelineTask.CustomRuns) != 1 {
-			return nil, resultRef.PipelineTask, fmt.Errorf("referenced tasks can only have length of 1 since a matrixed task does not support producing results, but was length %d", len(referencedPipelineTask.TaskRuns))
-		}
-		customRun := referencedPipelineTask.CustomRuns[0]
-		runName = customRun.GetObjectMeta().GetName()
-		runValue, err = findRunResultForParam(customRun, resultRef)
-		resultValue = *v1.NewStructuredValues(runValue)
-		if err != nil {
-			return nil, resultRef.PipelineTask, err
-		}
-	} else {
-		// Check to make sure the referenced task is not a matrix since a matrix does not support producing results
-		if len(referencedPipelineTask.TaskRuns) != 1 {
-			return nil, resultRef.PipelineTask, fmt.Errorf("referenced tasks can only have length of 1 since a matrixed task does not support producing results, but was length %d", len(referencedPipelineTask.TaskRuns))
-		}
-		taskRun := referencedPipelineTask.TaskRuns[0]
-		taskRunName = taskRun.Name
-		resultValue, err = findTaskResultForParam(taskRun, resultRef)
-		if err != nil {
-			return nil, resultRef.PipelineTask, err
-		}
+func resolveResultRef(taskRuns []*v1.TaskRun, resultRef *v1.ResultRef) (*ResolvedResultRef, error) {
+	taskRun := taskRuns[0]
+	taskRunName := taskRun.Name
+	resultValue, err := findTaskResultForParam(taskRun, resultRef)
+	if err != nil {
+		return nil, err
 	}
-
 	return &ResolvedResultRef{
 		Value:           resultValue,
 		FromTaskRun:     taskRunName,
-		FromRun:         runName,
+		FromRun:         "",
```

Comment:
Could you help me to understand why is this change?

Reply:
I separated `resolveCustomResultRef` and `ResolvedResultRef` into 2 functions so `ResolvedResultRef` would never have a `runName` since this is reserved for CustomRuns from my understanding.
https://github.com/tektoncd/pipeline/blob/8100f01ab5fee5d298e78565204c01744a76b8d7/pkg/reconciler/pipelinerun/resources/resultrefresolution.go#L155